File: apps/android/src/main/kotlin/space/bitos/app/ui/create/meme/MemeExportJobStore.kt
package space.bitos.app.ui.create.meme

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Durable export jobs (MUX-05): the rendered artifact is persisted BEFORE
 * the destination save, so an interrupted save retries from the artifact
 * without re-rendering. A job stuck in `saving` on relaunch becomes
 * needsReview — MediaStore saves cannot be reconciled back reliably, so
 * the honest recovery asks the user to check. Thread-safe (called from IO).
 */
class MemeExportJobStore(context: Context) {
    private val dir = File(context.filesDir, "meme-export-jobs").apply { runCatching { mkdirs() } }
    private val ledger = File(dir, "jobs.json")
    private val cap = 10

    data class Job(
        val id: Int, val format: String, val phase: String,
        val artifactFile: String, val artifactBytes: Int, val lastError: String?,
    )

    @Synchronized fun list(): List<Job> = runCatching {
        val array = JSONArray(ledger.readText())
        List(array.length()) { i ->
            val o = array.getJSONObject(i)
            val phase = o.getString("phase")
            Job(
                id = o.getInt("id"), format = o.getString("format"),
                // Crash-mid-save reconciliation: ask, don't assume.
                phase = if (phase == "saving") "needsReview" else phase,
                artifactFile = o.getString("artifactFile"),
                artifactBytes = o.optInt("artifactBytes"),
                lastError = o.optString("lastError").ifBlank { null },
            )
        }
    }.getOrDefault(emptyList())

    @Synchronized fun recoverable(): List<Job> =
        list().filter { it.phase == "rendered" || it.phase == "needsReview" || it.phase == "failed" }

    private fun persist(jobs: List<Job>) {
        runCatching {
            ledger.writeText(JSONArray().apply {
                jobs.forEach {
                    put(JSONObject().apply {
                        put("id", it.id); put("format", it.format); put("phase", it.phase)
                        put("artifactFile", it.artifactFile); put("artifactBytes", it.artifactBytes)
                        put("lastError", it.lastError ?: "")
                    })
                }
            }.toString())
        }
    }

    @Synchronized fun begin(format: String): Int {
        val id = ((System.currentTimeMillis() / 1000) % 10_000).toInt()
        val job = Job(id, format, "rendering", "export-$id.bin", 0, null)
        persist(listOf(job) + list().filterNot { it.id == id }.take(cap - 1))
        return id
    }

    /** Persist-then-effect: the artifact lands on disk before the save. */
    @Synchronized fun artifactReady(id: Int, bytes: ByteArray): Boolean {
        val jobs = list().toMutableList()
        val index = jobs.indexOfFirst { it.id == id }
        if (index < 0) return false
        val ok = runCatching { File(dir, jobs[index].artifactFile).writeBytes(bytes) }.isSuccess
        if (ok) jobs[index] = jobs[index].copy(phase = "rendered", artifactBytes = bytes.size)
        persist(jobs)
        return ok
    }

    @Synchronized fun update(id: Int, phase: String? = null, error: String? = null) {
        val jobs = list().toMutableList()
        val index = jobs.indexOfFirst { it.id == id }
        if (index < 0) return
        var job = jobs[index]
        phase?.let { job = job.copy(phase = it) }
        error?.let { job = job.copy(lastError = it) }
        jobs[index] = job
        persist(jobs)
    }

    @Synchronized fun finish(id: Int) {
        val jobs = list()
        jobs.firstOrNull { it.id == id }?.let { runCatching { File(dir, it.artifactFile).delete() } }
        persist(jobs.filterNot { it.id == id })
    }

    @Synchronized fun discard(id: Int) = finish(id)

    @Synchronized fun loadArtifact(id: Int): Pair<ByteArray, String>? {
        val job = list().firstOrNull { it.id == id } ?: return null
        val bytes = runCatching { File(dir, job.artifactFile).readBytes() }.getOrNull() ?: return null
        return bytes to job.format
    }
}

File: apps/ios/BitOS/Features/Create/MemeExportJobStore.swift
import Foundation

/// Durable export jobs (MUX-05): the rendered artifact is persisted BEFORE
/// the destination save, so an interrupted save retries from the artifact
/// without re-rendering. A job stuck in `saving` on relaunch becomes
/// needsReview — Photos saves are add-only and cannot be queried back, so
/// the honest reconciliation is asking the user to check.
struct MemeExportJob: Codable, Identifiable, Equatable {
    var v: Int = 1
    var id: Int
    var createdAtMs: Int64
    var updatedAtMs: Int64
    var format: String      // "png" | "gif" | "mp4"
    /// rendering | rendered | saving | needsReview | done
    var phase: String
    var artifactFile: String
    var artifactBytes: Int
    var lastError: String?
}

struct MemeExportJobStore {
    let directory: URL
    private(set) var jobs: [MemeExportJob] = []
    private let cap = 10

    init(directory: URL? = nil) {
        let base = directory ?? FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("meme-export-jobs", isDirectory: true)
        self.directory = base
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        if let data = try? Data(contentsOf: base.appendingPathComponent("jobs.json")),
           let decoded = try? JSONDecoder().decode([MemeExportJob].self, from: data) {
            jobs = decoded.map { job in
                // A crash mid-save can't be reconciled against Photos
                // (add-only) — surface it for review instead of assuming.
                job.phase == "saving" ? Self.phaseNeedsReview(job) : job
            }
        }
    }

    /// A crash mid-save can't be reconciled against Photos (add-only) —
    /// surface it for review instead of assuming success or failure.
    private static func phaseNeedsReview(_ job: MemeExportJob) -> MemeExportJob {
        var reviewed = job
        reviewed.phase = "needsReview"
        return reviewed
    }

    private func persist(_ list: [MemeExportJob]) {
        if let data = try? JSONEncoder().encode(list) {
            try? data.write(to: directory.appendingPathComponent("jobs.json"), options: .atomic)
        }
    }

    @discardableResult
    mutating func begin(format: String, nowMs: Int64) -> Int {
        let id = Int((nowMs / 1000) % 10_000)
        let job = MemeExportJob(
            id: id, createdAtMs: nowMs, updatedAtMs: nowMs,
            format: format, phase: "rendering",
            artifactFile: "export-\(id)-\(nowMs).bin", artifactBytes: 0
        )
        jobs.removeAll { $0.id == id }
        jobs.insert(job, at: 0)
        jobs = Array(jobs.prefix(cap))
        persist(jobs)
        return id
    }

    /// Artifact persisted BEFORE the destination save (persist-then-effect).
    mutating func artifactReady(_ id: Int, bytes: Data, nowMs: Int64) -> Bool {
        guard let index = jobs.firstIndex(where: { $0.id == id }) else { return false }
        let ok = (try? bytes.write(to: directory.appendingPathComponent(jobs[index].artifactFile), options: .atomic)) != nil
        if ok {
            jobs[index].phase = "rendered"
            jobs[index].artifactBytes = bytes.count
        }
        jobs[index].updatedAtMs = nowMs
        persist(jobs)
        return ok
    }

    mutating func update(_ id: Int, phase: String? = nil, error: String? = nil, clearError: Bool = false, nowMs: Int64) {
        guard let index = jobs.firstIndex(where: { $0.id == id }) else { return }
        if let phase { jobs[index].phase = phase }
        if let error { jobs[index].lastError = error }
        if clearError { jobs[index].lastError = nil }
        jobs[index].updatedAtMs = nowMs
        persist(jobs)
    }

    /// Done: the destination holds the output — the artifact goes away.
    mutating func finish(_ id: Int) {
        guard let index = jobs.firstIndex(where: { $0.id == id }) else { return }
        try? FileManager.default.removeItem(at: directory.appendingPathComponent(jobs[index].artifactFile))
        jobs.remove(at: index)
        persist(jobs)
    }

    mutating func discard(_ id: Int) {
        guard let job = jobs.first(where: { $0.id == id }) else { return }
        try? FileManager.default.removeItem(at: directory.appendingPathComponent(job.artifactFile))
        jobs.removeAll { $0.id == id }
        persist(jobs)
    }

    func job(_ id: Int) -> MemeExportJob? { jobs.first { $0.id == id } }
    func loadArtifact(_ job: MemeExportJob) -> Data? {
        try? Data(contentsOf: directory.appendingPathComponent(job.artifactFile))
    }
    /// Jobs whose rendered artifact is intact (retry reuses it, no rerender).
    var recoverable: [MemeExportJob] { jobs.filter { $0.phase == "rendered" || $0.phase == "needsReview" || $0.phase == "failed" } }
}
